api_callable.go

```diff
@@ -0,0 +1,98 @@
+package gax
+
+import (
+	"golang.org/x/net/context"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"time"
+)
+
+var (
+	retryableGrpcCodes []codes.Code = []codes.Code{
+		codes.DeadlineExceeded,
+		codes.Aborted,
+	}
+)
+
+// Represents a GRPC call stub.
+type APICall func(context.Context, interface{}) (interface{}, error)
+
+// isRetryable returns true if c represents a retryable GRPC code.
+func isRetryable(c codes.Code) bool {
+	for _, code := range retryableGrpcCodes {
+		if c == code {
+			return true
+		}
+	}
+	return false
+}
+
+// shorter returns the shorter time.Duration between a and b.
+func shorter(a time.Duration, b time.Duration) time.Duration {
+	if a < b {
+		return a
+	}
+	return b
+}
+
+// scaleDuration returns a copy of a * mult.
+func scaleDuration(a time.Duration, mult float64) time.Duration {
+	ns := float64(a) * mult
+	return time.Duration(int64(ns))
+}
+
+// stubWithRetry returns a wrapper for stub with a exponential backoff retry
+// mechanism based on the values provided in backoffSettings.
+func stubWithRetry(ctx context.Context, stub APICall, backoffSettings BackoffSettings) APICall {
+	return func(ctx context.Context, req interface{}) (resp interface{}, err error) {
+		// Forces ctx to expire after a deadline.
+		ctx, _ = context.WithTimeout(ctx, backoffSettings.TotalTimeoutMillis)
+
+		delay := backoffSettings.InitialRetryDelayMillis
+		timeout := backoffSettings.InitialRPCTimeoutMillis
+
+		for {
+			// If the deadline is exceeded...
+			if ctx.Err() != nil {
+				return nil, ctx.Err()
+			}
+			timeoutCtx, _ := context.WithTimeout(ctx, timeout)
+			resp, err = stub(timeoutCtx, req)
+			code := grpc.Code(err)
+			if code == codes.OK {
+				break
+			}
+			if !isRetryable(code) {
+				return nil, err
+			}
+			delayCtx, _ := context.WithTimeout(ctx, delay)
+			<-delayCtx.Done()
+
+			delay = shorter(scaleDuration(delay, backoffSettings.RetryDelayMultiplier),
+				backoffSettings.MaxRetryDelayMillis)
+			timeout = shorter(scaleDuration(timeout, backoffSettings.RPCTimeoutMultiplier),
+				backoffSettings.MaxRPCTimeoutMillis)
```

Comment:
You don't have to check which is longer for delay/timeout or max timeouts.
A context will finish with timeout if the parent context is timed-out, or it is timed-out (like this: http://play.golang.org/p/YJng0EZXnz -- child_ctx.Done() ends in 1 second).

Because total timeout is set at line 49, delayCtx and timeoutCtx never pass the total timeout. You can simply scale them with multiplier.